SU2_CFD/include/solvers/CSolverFactory.hpp

```diff
@@ -0,0 +1,186 @@
+ *
+ * SU2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include "../../../Common/include/option_structure.hpp"
+
+/*!
+ * \brief Enum of different sub solvers the main solver can use. There is not a 1-to-1 correspondence between the actual classes
+ * and the types of sub solvers, as one class can be used for several sub solvers.
+ */
+enum class SUB_SOLVER_TYPE {
+  CONT_ADJ_EULER,          /*!< \brief Continuous Adjoint Euler solver  */
+  CONT_ADJ_NAVIER_STOKES,  /*!< \brief Continuous Adjoint Navier Stokes solver  */
+  CONT_ADJ_TURB,           /*!< \brief Continuous Adjoint Turbulent solver  */
+  BASELINE,                /*!< \brief Baseline solver  */
+  TEMPLATE,                /*!< \brief Template solver  */
+  BASELINE_FEM,            /*!< \brief Baseline FEM solver */
+  DISC_ADJ_FEA,            /*!< \brief Discrete adjoint FEA solver  */
+  DISC_ADJ_MESH,           /*!< \brief Discrete adjoint mesh solver */
+  DISC_ADJ_FLOW,           /*!< \brief Discrete adjoint flow solver */
+  DISC_ADJ_TURB,           /*!< \brief Discrete adjoint turbulence solver */
+  DISC_ADJ_HEAT,           /*!< \brief Discrete adjoint heat solver */
+  EULER,                   /*!< \brief Compressible Euler solver */
+  NAVIER_STOKES,           /*!< \brief Compressible Navier-Stokes solver */
+  INC_EULER,               /*!< \brief Incompressible Euler solver */
+  INC_NAVIER_STOKES,       /*!< \brief Incompressible Navier-stokes solver */
+  FEA,                     /*!< \brief Structural Finite-Element solver */
+  DG_EULER,                /*!< \brief Higher-order DG Euler solver*/
+  DG_NAVIER_STOKES,        /*!< \brief Higher-order DG Navier-Stokes solver*/
+  HEAT,                    /*!< \brief Heat solver */
+  TRANSITION,              /*!< \brief Transition model solver*/
+  TURB_SA,                 /*!< \brief SA turbulence model solver */
+  TURB_SST,                /*!< \brief SST turbulence model solver */
+  TURB,                    /*!< \brief Turbulence model solver */
+  MESH,                    /*!< \brief Mesh solver */
+  RADIATION,               /*!< \brief Radiation solver */
+  DISC_ADJ_RADIATION,      /*!< \brief Discrete adjoint radiation solver */
+  NONE
+};
+
+enum class INTEGRATION_TYPE{
+  MULTIGRID,
+  SINGLEGRID,
+  DEFAULT,
+  FEM_DG,
+  STRUCTURAL,
+  NONE
+};
+
+struct SolverMetaData{
+  SUB_SOLVER_TYPE  solverType        = SUB_SOLVER_TYPE::NONE;
+  INTEGRATION_TYPE integrationType   = INTEGRATION_TYPE::NONE;
+};
+
+class CSolver;
+class CGeometry;
+class CConfig;
+
+class CSolverFactory {
+  
+private:
+
+  static map<CSolver*, SolverMetaData> allocatedSolvers;
```

Comment:
Cool! I like this even more, only one place to edit, can you make the key (CSolver*) const? Just to make sure this does not get misused.
Not so sure on the numerics idea though, isn't that going to make the solver factory too complex? (there are a lot more numerics options per solver than there are integration options).

Reply:
Yes that might mean that it gets more complex. But then everything the solver needs is determined at one place. I kind of like that idea.